sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala

```diff
@@ -250,11 +281,15 @@ case class BroadcastQueryStageExec(
     reuse
   }
 
-  override def cancel(): Unit = {
+  override def cancel(): StageCancellationStatus = {
     if (!broadcast.relationFuture.isDone) {
```

Comment:
shall we do something like shuffle query stage ?

Reply:
`BroadcastExchangeExec.relationFuture` returns Java Future but ShuffleExchangeExec.mapOutputStatisticsFuture returns Scala Future. However, i reproduced today as `BroadcastQueryStageExec` can also have same problem although it has this check(`!broadcast.relationFuture.isDone`). For example, we can also have following case:

**Before fix:**
```
BroadcastQueryStage1.doMaterialize is called and it is failed,
BroadcastQueryStage1 is excluded from cancellation due to earlyFailedStage,
BroadcastQueryStage2 is tried to cancel,
BroadcastQueryStage2.relationFuture needs to be initialized because it is not materialized yet,
BroadcastQueryStage2.relationFuture cancellation is finished with COMPLETED status,
Same for other BroadcastQueryStages which are not materialized yet.
```
**After fix:**
```
BroadcastQueryStage1.doMaterialize is called and it is failed,
BroadcastQueryStage1 is excluded from cancellation due to earlyFailedStage,
BroadcastQueryStage2 is tried to cancel,
BroadcastQueryStage2.relationFuture cancellation is skipped with SKIPPED status because it is not materialized yet,
Same for other BroadcastQueryStages which are not materialized yet.
```